src/transformers/pipelines/automatic_mask_generation.py

```diff
@@ -0,0 +1,615 @@
+        coords = coords.reshape(-1, 4)
+
+    return coords
+
+
+def _generate_crop_boxes(
+    image,
+    target_size,
+    n_layers: int = 0,
+    overlap_ratio: float = 512 / 1500,
+    points_per_crop: Optional[int] = 32,
+    scale_per_layer: Optional[List[int]] = 1,
+) -> Tuple[List[List[int]], List[int]]:
+    """
+    Generates a list of crop boxes of different sizes. Each layer has (2**i)**2 boxes for the ith layer.
+    """
+    if isinstance(image, list):
+        raise ValueError("Only one image is allowed for crop generation.")
+    image = to_numpy_array(image)
+    original_size = image.shape[:2]
+
+    points_grid = []
+    for i in range(n_layers + 1):
+        n_points = int(points_per_crop / (scale_per_layer**i))
+        points_grid.append(_build_point_grid(n_points))
+
+    crop_boxes, layer_idxs = [], []
+    im_h, im_w = original_size
+    short_side = min(im_h, im_w)
+
+    # Original image
+    crop_boxes.append([0, 0, im_w, im_h])
+    layer_idxs.append(0)
+
+    for i_layer in range(n_layers):
+        n_crops_per_side = 2 ** (i_layer + 1)
+        overlap = int(overlap_ratio * short_side * (2 / n_crops_per_side))
+
+        crop_w = int(math.ceil((overlap * (n_crops_per_side - 1) + im_w) / n_crops_per_side))
+        crop_h = int(math.ceil((overlap * (n_crops_per_side - 1) + im_h) / n_crops_per_side))
+
+        crop_box_x0 = [int((crop_w - overlap) * i) for i in range(n_crops_per_side)]
+        crop_box_y0 = [int((crop_h - overlap) * i) for i in range(n_crops_per_side)]
+
+        # Crops in XYWH format
+        for x0, y0 in product(crop_box_x0, crop_box_y0):
+            box = [x0, y0, min(x0 + crop_w, im_w), min(y0 + crop_h, im_h)]
+            crop_boxes.append(box)
+            layer_idxs.append(i_layer + 1)
+
+    # generate cropped images
+    cropped_images = []
+    total_points_per_crop = []
+    for i, crop_box in enumerate(crop_boxes):
+        x0, y0, x1, y1 = crop_box
+        cropped_im = image[y0:y1, x0:x1, :]
+        cropped_images.append(cropped_im)
+
+        cropped_im_size = cropped_im.shape[:2]
+        points_scale = np.array(cropped_im_size)[None, ::-1]
+
+        total_points_per_crop.append(points_grid[layer_idxs[i]] * points_scale)
+
+    normalized_total_points_per_crop = []
+    for points_per_crop in total_points_per_crop:
+        normalized_total_points_per_crop.append(
+            [_normalize_coordinates(target_size, point, original_size) for point in points_per_crop]
+        )
+
+    crop_boxes = torch.tensor(crop_boxes, dtype=torch.float32)
+    points_per_crop = torch.stack([torch.tensor([p]) for p in np.array(normalized_total_points_per_crop)], dim=0)
+    points_per_crop = points_per_crop.permute(0, 2, 1, 3)
+
+    input_labels = torch.ones_like(points_per_crop[:, :, :, 0], dtype=torch.long)
+
+    return crop_boxes, points_per_crop, cropped_images, input_labels
+
+
+def _uncrop_masks(masks, crop_box: List[int], orig_h: int, orig_w: int):
```

Comment:
A lot seems copied from the SAM PR here. Is there a way to use a method in the processor instead of duplicating the code?

Reply:
they were removed from the SAM PR